vunit/sim_if/xcelium.py

```diff
@@ -0,0 +1,694 @@
+from ..color_printer import NO_COLOR_PRINTER
+from ..configuration import Configuration
+from ..project import Project
+from vunit.source_file import SourceFile
+from vunit.library import Library
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class XceliumInterface(  # pylint: disable=too-many-instance-attributes
+    SimulatorInterface
+):
+    """
+    Interface for the Cadence Xcelium simulator
+    """
+
+    name = "xcelium"
+    supports_gui_flag = True
+    package_users_depend_on_bodies = False
+    incremental = False
+
+    compile_options = [
+        ListOfStringOption("xcelium.xrun_vhdl_flags"),
+        ListOfStringOption("xcelium.xrun_verilog_flags"),
+        ListOfStringOption("xcelium.xrun_cdslib"),
+    ]
+
+    sim_options = [ListOfStringOption("xcelium.xrun_sim_flags"),
+                   ListOfStringOption("xcelium.xrun_sim_flags.gui")]
+
+    _global_compile_options: Dict
+
+    @staticmethod
+    def add_arguments(parser):
+        """
+        Add command line arguments
+        """
+        group = parser.add_argument_group(
+            "Xcelium/Incisive", description="Xcelium/Incisive specific flags"
+        )
+        group.add_argument(
+            "--cdslib",
+            default=None,
+            help="The cds.lib file to use. If not given, VUnit maintains its own cds.lib file.",
+        )
+        group.add_argument(
+            "--hdlvar",
+            default=None,
+            help="The hdl.var file to use. If not given, VUnit does not use a hdl.var file.",
+        )
+        group.add_argument(
+            "--use_indago",
+            default=False,
+            help="Indicate whether to use Indago instead of Simvision when running in GUI mode.",
+            action="store_true",
+        )
+
+    @classmethod
+    def from_args(cls, args, output_path, **kwargs):
+        """
+        Create new instance from command line arguments object
+        """
+        return cls(
+            prefix=cls.find_prefix(),
+            output_path=output_path,
+            log_level=args.log_level,
+            gui=args.gui,
+            cdslib=args.cdslib,
+            hdlvar=args.hdlvar,
+            use_indago=args.use_indago
+        )
+
+    @classmethod
+    def find_prefix_from_path(cls):
+        """
+        Find xcelium simulator from PATH environment variable
+        """
+        return cls.find_toolchain(["xrun"])
+
+
+    @staticmethod
+    def supports_vhdl_contexts():
+        """
+        Returns True when this simulator supports VHDL 2008 contexts
+        """
+        return True
+
+    def __init__(  # pylint: disable=too-many-arguments
+        self, prefix, output_path, gui=False, log_level=None, cdslib=None, hdlvar=None, use_indago=False
+    ):
+        SimulatorInterface.__init__(self, output_path, gui)
+        self._global_compile_options = {}
+        self._prefix = prefix
+        self._libraries = []
+        self._log_level = log_level
+        if cdslib is None:
```

Comment:
Why check for None instead of doing:

```
if cdslib:
   ...
else:
   ...
```

Shorter if it fits reasonably in VUnit's line length limit (if there is one):

```
x = a if cdslib else b
```

Reply:
This is actually as per https://peps.python.org/pep-0008/#programming-recommendations, but it could be that an unset `cdslib` should have a default of `""` which would then be tested as you show.

Reply:
Well yeah, but only if you need to know that the value is actually `None`. However, all that is being checked here is if the value is set, otherwise it's being set to the default.

In this case, it is written as an anti-pattern. My comment here is implicitly recommends avoiding this pattern by using the more logical pattern of handling `None` and other "empty" objects:

```
if not value:
  do this
else:
  do that
```

Why would you check for "false" first if you can just do the following which is less confusing to read (more important for complicated conditionals):

```
if value:
  do that
else:
  do this
```

`None` is falsy meaning you can safely use `if [str, None]` and it will never be true for `None`. The other advantage of not checking `None` here is that it will also catch empty objects like `""`, `[]`, `()`, `{}`, etc. and fallback to the default behavior. As it is written, those objects will require the user to read through a code traceback when it inevitably fails because the user used the wrong empty type. I think it's important here because there's no type hint, so I'm not actually sure what `cdslib` is supposed to be. 

Regarding the PEPs warning:

>Also, beware of writing if x when you really mean if x is not None – e.g. when testing whether a variable or argument >that defaults to None was set to some other value. The other value might have a type (such as a container) that could >be false in a boolean context!

It doesn't recommend one way or another. However, I use type-hints for this reason so the user knows what is expected and what isn't. In that case, it is safe to make assumptions about how the data can be handled as the interfaces requirements are explicit.

Reply:
`cdslib` defaults to `None`, this is the correct check. In terms of type-hinting, we didn't feel it was our place to universally add type-hinting throughout all of VUnit as much as it pained us to not do so.

Reply:
I should also clarify that `""` is a valid input for this field. If the user does not specify, or does not wish to specify the `cds.lib` file, then Xcelium will infer the location of the file and handle it automatically.

Reply:
>In terms of type-hinting, we didn't feel it was our place to universally add type-hinting throughout all of VUnit as much as it pained us to not do so.

I totally understand that. 

`""`, at least to me, is not obvious that it's supposed to be CWD unless I knew the implementation. Yes, `Path("").resolve()` resolves to the CWD, but this to me isn't clear. I would think that if a user wanted CWD, they would pass `"."`. If I were reading the code and `""` was used, I would read it as the user letting the code choose the default. My preference is to be explicit, not implicit and in this case I think treating `""` as a valid path is implicit.